compaction/compaction_manager.cc

```diff
@@ -343,6 +359,7 @@ future<sstables::compaction_result> compaction_manager::task::compact_sstables(s
             release_exhausted(old_sstables);
         }
     };
+    descriptor.retrieve_owned_ranges_if_required(_cm.get_compaction_state(_compacting_table));
 
```

Comment:
Like I said above, I think we can bypass this if descriptor already has owned_ranges_ptr set.

Reply:
This is on purpose since I want the `bool requires_cleanup` flag to be authoritative so we don't blindly do the cleanup any time the owned_ranges is set, as it's set also in other cases than cleanup, and the "if_required" part of the function name is for that purpose.  Maybe the function name is too specific since it's doing more than just retrieving owned_ranges. It marks the descriptor for cleanup as well as retrieving the the owned_ranges,  if required.

Reply:
Before I went to sleep last night, I thought about this and here is my reasoning: We're moving towards a scenario where compaction types can all perform cleanup if it has information about how to do so. What's needed for a compaction to filter out a token that is no longer owned by a node? Set of owned ranges, right? Will there ever be a case where we **need** to set owned_ranges_ptr and say to not cleanup, or vice versa? Don't think so. It's not needed. Compaction manager communicates its intention of cleanup to compaciton job via setting the owned_ranges_ptr.

This is exactly the same logic we applied to tombstone purge. If we have sstable set snapshot, it means that compaction will be able to purge expired tombstones. Resharding on boot doesn't have sstable set snapshot by the time it runs, therefore it doesn't enable the purging. Why would we set the snapshot and tell compaction to not purge tombstones if possible? It's redundant to both set snapshot and mark the descriptor as requiring tombstone purge. We don't need a flag for that. The information is already there.

My suggestion is to add compaction_descriptor::enable_token_cleanup(owned_ranges_ptr), then compaction_manager can easily enable cleanup when building the descriptor. I think it will be very good this way.

Reply:
OK, I concur.
I can also refactor the code differently so that the decision would be done at the compaction_manager layer and the descriptor logic will remain more basic.

Reply:
done in v4